### ISSUE TITLE
validation loss (rpn_bbox_loss) is very high 

### ISSUE THREAD

**Issue by geobowen** [matterport/Mask_RCNN#2261]
I used the Mask_RCNN to train my own dataset which has 3 classes. The initial experiment results seem ok but I want to improve it so I add more epoches. I have 719 training data and use data augmentation during the training.

Config：

    GPU_COUNT = 1
    IMAGES_PER_GPU = 1
 
    NUM_CLASSES = 1 + 3  # background + 3 shapes
 
    IMAGE_MIN_DIM = 512
    IMAGE_MAX_DIM = 512

    RPN_ANCHOR_SCALES = (16,32,64,128,256)  # anchor side in pixels
    
    TRAIN_ROIS_PER_IMAGE = 200
 
    STEPS_PER_EPOCH = 240
 
    VALIDATION_STEPS = 30

I tried twice and training strategies are like:

**1st(blue loss graph)**
 head layers with lr=0.002 for 2 epoches + all layers with lr=0.001 for 38 epoches

**2nd(green loss graph)**
head layers with lr=0.002 for 2 epoches + all layers with lr=0.001 for 40 epoches + all layers with lr=0.0001 for 15 epoches

Both used pre-trained weights `mask_rcnn_coco.h5` and other parameters remain the same

**This is the training loss**
![image](https://user-images.githubusercontent.com/33017253/85939297-b7c1c480-b946-11ea-80d4-f3206ead2472.png)

**val loss**
![image](https://user-images.githubusercontent.com/33017253/85939326-ef307100-b946-11ea-91db-ed9c3e232078.png)

I just dont know why the val loss increased that much. And I think it is the `rpn_bbox_loss` that caused the great change and I found that the validation loss goes really high after the **head layers training was done**.
![image](https://user-images.githubusercontent.com/33017253/85939687-34a26d80-b94a-11ea-9e0d-022f54f60eaf.png)
![image](https://user-images.githubusercontent.com/33017253/85939963-205f7000-b94c-11ea-8861-8597b4c5901f.png)

 I tried many ways to solve the problem and I checked the validation data again. I think the masks are loading correctly and i tried more `STEPS_PER_EPOCH=719` to match my training dataset and the problem happened every time.

Does anyone know which part might go wrong and help me solve the problem. I appreciate it a lot

PS **I recreate the anaconda environment** before i started the second training because something is wrong with the IPython

Best Regards

**Comment by konstantin-frolov** [comment]
Rpn box loss - loss how accurate you detecting ROI’s for next classification and segmentation. If you work with small or big objects try change RPN_ANCHORS_SCALE. 
And you learning rate high. Try learn only heads with lr<=0.001 about 20 epochs. Then 3+ layers - lr=1e-4, epochs = 50, 4+ - lr=3e-5, epochs= 80, all - lr=1e-5, epochs = 120

**Comment by kimile599** [comment]
> I used the Mask_RCNN to train my own dataset which has 3 classes. The initial experiment results seem ok but I want to improve it so I add more epoches. I have 719 training data and use data augmentation during the training.
> 
> Config：
> 
> ```
> GPU_COUNT = 1
> IMAGES_PER_GPU = 1
> 
> NUM_CLASSES = 1 + 3  # background + 3 shapes
> 
> IMAGE_MIN_DIM = 512
> IMAGE_MAX_DIM = 512
> 
> RPN_ANCHOR_SCALES = (16,32,64,128,256)  # anchor side in pixels
> 
> TRAIN_ROIS_PER_IMAGE = 200
> 
> STEPS_PER_EPOCH = 240
> 
> VALIDATION_STEPS = 30
> ```
> 
> I tried twice and training strategies are like:
> 
> **1st(blue loss graph)**
> head layers with lr=0.002 for 2 epoches + all layers with lr=0.001 for 38 epoches
> 
> **2nd(green loss graph)**
> head layers with lr=0.002 for 2 epoches + all layers with lr=0.001 for 40 epoches + all layers with lr=0.0001 for 15 epoches
> 
> Both used pre-trained weights `mask_rcnn_coco.h5` and other parameters remain the same
> 
> **This is the training loss**
> ![image](https://user-images.githubusercontent.com/33017253/85939297-b7c1c480-b946-11ea-80d4-f3206ead2472.png)
> 
> **val loss**
> ![image](https://user-images.githubusercontent.com/33017253/85939326-ef307100-b946-11ea-91db-ed9c3e232078.png)
> 
> I just dont know why the val loss increased that much. And I think it is the `rpn_bbox_loss` that caused the great change and I found that the validation loss goes really high after the **head layers training was done**.
> ![image](https://user-images.githubusercontent.com/33017253/85939687-34a26d80-b94a-11ea-9e0d-022f54f60eaf.png)
> ![image](https://user-images.githubusercontent.com/33017253/85939963-205f7000-b94c-11ea-8861-8597b4c5901f.png)
> 
> I tried many ways to solve the problem and I checked the validation data again. I think the masks are loading correctly and i tried more `STEPS_PER_EPOCH=719` to match my training dataset and the problem happened every time.
> 
> Does anyone know which part might go wrong and help me solve the problem. I appreciate it a lot
> 
> PS **I recreate the anaconda environment** before i started the second training because something is wrong with the IPython
> 
> Best Regards

I think we are running into same issues. The total loss seems ok, but val_loss is not converging at all. 
One more question, are you detecting long and thin objects?

**Comment by kimile599** [comment]
> Rpn box loss - loss how accurate you detecting ROI’s for next classification and segmentation. If you work with small or big objects try change RPN_ANCHORS_SCALE.
> And you learning rate high. Try learn only heads with lr<=0.001 about 20 epochs. Then 3+ layers - lr=1e-4, epochs = 50, 4+ - lr=3e-5, epochs= 80, all - lr=1e-5, epochs = 120

Thank you for the detailed suggestion for traiing steps. I am new to the mask rcnn. I am wondering, if there is a path to estimate how the parameters go with a small dataset like 200 pics. I don't want to label 2000 pictures right now and then start to trian the model.

**Comment by pullulation** [comment]
> > I used the Mask_RCNN to train my own dataset which has 3 classes. The initial experiment results seem ok but I want to improve it so I add more epoches. I have 719 training data and use data augmentation during the training.
> > Config：
> > ```
> > GPU_COUNT = 1
> > IMAGES_PER_GPU = 1
> > 
> > NUM_CLASSES = 1 + 3  # background + 3 shapes
> > 
> > IMAGE_MIN_DIM = 512
> > IMAGE_MAX_DIM = 512
> > 
> > RPN_ANCHOR_SCALES = (16,32,64,128,256)  # anchor side in pixels
> > 
> > TRAIN_ROIS_PER_IMAGE = 200
> > 
> > STEPS_PER_EPOCH = 240
> > 
> > VALIDATION_STEPS = 30
> > ```
> > 
> > 
> > I tried twice and training strategies are like:
> > **1st(blue loss graph)**
> > head layers with lr=0.002 for 2 epoches + all layers with lr=0.001 for 38 epoches
> > **2nd(green loss graph)**
> > head layers with lr=0.002 for 2 epoches + all layers with lr=0.001 for 40 epoches + all layers with lr=0.0001 for 15 epoches
> > Both used pre-trained weights `mask_rcnn_coco.h5` and other parameters remain the same
> > **This is the training loss**
> > ![image](https://user-images.githubusercontent.com/33017253/85939297-b7c1c480-b946-11ea-80d4-f3206ead2472.png)
> > **val loss**
> > ![image](https://user-images.githubusercontent.com/33017253/85939326-ef307100-b946-11ea-91db-ed9c3e232078.png)
> > I just dont know why the val loss increased that much. And I think it is the `rpn_bbox_loss` that caused the great change and I found that the validation loss goes really high after the **head layers training was done**.
> > ![image](https://user-images.githubusercontent.com/33017253/85939687-34a26d80-b94a-11ea-9e0d-022f54f60eaf.png)
> > ![image](https://user-images.githubusercontent.com/33017253/85939963-205f7000-b94c-11ea-8861-8597b4c5901f.png)
> > I tried many ways to solve the problem and I checked the validation data again. I think the masks are loading correctly and i tried more `STEPS_PER_EPOCH=719` to match my training dataset and the problem happened every time.
> > Does anyone know which part might go wrong and help me solve the problem. I appreciate it a lot
> > PS **I recreate the anaconda environment** before i started the second training because something is wrong with the IPython
> > Best Regards
> 
> I think we are running into same issues. The total loss seems ok, but val_loss is not converging at all.
> One more question, are you detecting long and thin objects?

**Comment by vyadava** [comment]
@pullulation  I am also facing same issue. Please can you tell me how did you solve the issue?

**Comment by putrasto** [comment]
I got exactly the same issue. The "loss" and "val_loss"  of "all" layers training become high after "heads" training.
In my case, the problem is the library "imgaug==0.4.0". I noticed this because on "heads" training the augmentation is None.
I suggest to use this version "imgaug==0.2.9"

**Comment by Kingofsunwang** [comment]
Hi, I am trying to draw loss graph but I do not know how to achieve it. Although this issue is not pertinent to my problem, can you explain how to draw if possible? Thank you so much.

**Comment by Dcharless** [comment]
Hey @Kingofsunwang did you find how to make the plots? i am having the same problem
